tensorflow/core/kernels/fused_batch_norm_op.cc

```diff
@@ -584,10 +729,29 @@ class FusedBatchNormOp : public OpKernel {
     OP_REQUIRES_OK(context, context->allocate_output(4, scale.shape(),
                                                      &saved_maybe_inv_var));
 
-    functor::FusedBatchNorm<Device, T, U>()(
-        context, x, scale, offset, estimated_mean, estimated_variance, epsilon_,
-        y, batch_mean, batch_var, saved_mean, saved_maybe_inv_var,
-        tensor_format_, is_training_);
+    if (!use_reserved_space) {
+      functor::FusedBatchNorm<Device, T, U>()(
+          context, x, scale, offset, estimated_mean, estimated_variance,
+          epsilon_, y, batch_mean, batch_var, saved_mean, saved_maybe_inv_var,
+          tensor_format_, nullptr, nullptr, is_training_);
+    } else {
+#if GOOGLE_CUDA
+      functor::CudnnBatchNormAllocatorInOutput<U> reserve_space_allocator(
+          context, 5);
+      functor::CudnnBatchNormAllocatorInTemp<uint8> workspace_allocator(
+          context);
+#else
+      // Allocate two dummy objects to handle the 5th output. If they are seen
+      // by the CPU functor, we will allocate empty space for the 5th output.
+      Tensor reserve_space_allocator;
+      Tensor workspace_allocator;
```

Comment:
I would rather create a dummy CudnnBatchNormAllocatorInTemp that does nothing and a CudnnBatchNormAllocatorInOutput that allocates 5th output in destructor, when GOOGLE_CUDA is 0, instead of casting a tensor pointer to a allocator pointer.

That can also help to eliminate the `#if GOOGLE_CUDA` here.

Reply:
Same reason of the above. CudnnBatchNormAllocatorInTemp inherits from ScratchAllocator which is defined in stream_executor.h. And stream_executor.h may not be found in non-GPU environment.